pkg/controller/ocsinitialization/ocsinitialization_controller.go

```diff
@@ -0,0 +1,188 @@
+}
+
+// blank assignment to verify that ReconcileOCSInitialization implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileOCSInitialization{}
+
+// ReconcileOCSInitialization reconciles a OCSInitialization object
+type ReconcileOCSInitialization struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a OCSInitialization object and makes changes based on the state read
+// and what is in the OCSInitialization.Spec
+// TODO(user): Modify this Reconcile function to implement your Controller logic.  This example creates
+// a Pod as an example
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileOCSInitialization) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling OCSInitialization")
+
+	initNamespacedName := InitNamespacedName()
+	instance := &ocsv1alpha1.OCSInitialization{}
+	if initNamespacedName.Name != request.Name || initNamespacedName.Namespace != request.Namespace {
+		// Ignoring this resource because it has the wrong name or namespace
+		reqLogger.Info(wrongNamespacedName)
+		err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+		if err != nil {
+			// the resource probably got deleted
+			if errors.IsNotFound(err) {
+				return reconcile.Result{}, nil
+			}
+			return reconcile.Result{}, err
+		}
+		instance.Status.ErrorMessage = wrongNamespacedName
+
+		err = r.client.Status().Update(context.TODO(), instance)
+		if err != nil {
+			reqLogger.Error(err, "failed to update ignored resource")
+		}
+		return reconcile.Result{}, err
+	}
+
+	// Fetch the OCSInitialization instance
+	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Recreating since we depend on this to exist. A user may delete it to
+			// induce a reset of all initial data.
+			reqLogger.Info("recreating OCSInitialization resource")
+			return reconcile.Result{}, r.client.Create(context.TODO(), &ocsv1alpha1.OCSInitialization{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      initNamespacedName.Name,
+					Namespace: initNamespacedName.Namespace,
+				},
+			})
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	if instance.Status.StorageClassesCreated == true {
+		// we only create the data once and then allow changes or even deletion, so we
```

Comment:
What if the user makes changes to the CR Spec?

There should be a single source of truth for StorageClass configuration. Given that they are initially configured here in a CR, updates to the CR should be applied to the cluster. Additionally, direct updates to the cluster should be detected and reverted as per CR.